Src/VimCore/WordUtil.fs

```diff
@@ -1,23 +1,97 @@
+
+    member x.KeywordCharSet = _keywordCharSet
+    member x.IsBigWordChar c = not (Char.IsWhiteSpace(c))
+    member x.IsNormalWordChar c = _keywordCharSet.Contains c
+    member x.IsBigWordOnlyChar c = (not (x.IsNormalWordChar c)) && (not (Char.IsWhiteSpace(c)))
+
+    member x.IsWordChar wordKind c =
+        match wordKind with
+        | WordKind.BigWord -> x.IsBigWordChar c
+        | WordKind.NormalWord -> x.IsNormalWordChar c 
+
+    /// Get the word predicate given the current char and word wordKind. The returned predicate 
+    /// will match all valid characters in the word based on the initial char. 
+    member x.GetWordPredicateNonWhitespace wordKind c =
+        Debug.Assert(not (CharUtil.IsWhiteSpace c))
+        match wordKind with
+        | WordKind.NormalWord -> 
+            if x.IsNormalWordChar c then x.IsNormalWordChar
+            else x.IsBigWordOnlyChar
+        | WordKind.BigWord -> x.IsBigWordChar
+
+    member x.GetWordPredicate wordKind c =
+        if CharUtil.IsWhiteSpace c then None
+        else x.GetWordPredicateNonWhitespace wordKind c |> Some
+
+    /// Get the word spans on the text in the given direction
+    member x.GetWordSpansInText wordKind path (text: string) =
+        // Build up a sequence to get the words in the line
+        let wordsForward = 
+            0
+            |> Seq.unfold (fun index ->
+                // Get the start index of the word and the predicate to keep matching
+                // the word
+                let rec getWord index = 
+                    if index < text.Length then
+                        let lineBreakLength = EditUtil.GetLineBreakLength text index
+                        if lineBreakLength = 0 then
+                            match x.GetWordPredicate wordKind text.[index] with
+                            | Some predicate -> 
+                                let mutable endIndex = index + 1
+                                while endIndex < text.Length && predicate text.[endIndex] do
+                                    endIndex <- endIndex + 1
+                                Some (Span.FromBounds(index, endIndex), endIndex)
+                            | None -> 
+                                // Go to the next index
+                                getWord (index + 1)
+                        elif index > 0 && EditUtil.IsInsideLineBreak text (index - 1) then
+                            // Empty line is a word
+                            let endIndex = index + lineBreakLength
+                            Some (Span.FromBounds(index, endIndex), endIndex)
+                        else
+                            getWord (index + lineBreakLength)
+                    else
+                        None
+                getWord index)
+
+        // Now return the actual sequence 
+        match path with
+        | SearchPath.Forward -> wordsForward
+        | SearchPath.Backward -> wordsForward |> List.ofSeq |> List.rev |> Seq.ofList

```

Comment:
There isn't actually a `Seq.rev` function because implementing it requires the entire sequence to be evaluated and cached to return the first element. Not much of a sequence at that point which is why I assume it wasn't added. Have to put it into a `List` first then reverse it. 

That being said this is still rather inefficient. Wonder if it's just easier to write the reverse search. Let me see.

Reply:
Went ahead and implemented `SearchPath.Backward` in the same way as `SearchPath.Forward` by using `Seq.unfold`. Glad I did this because it forced me to re-examine other parts of the code here and as a result I made two changes:

1. `GetFullWordSpanInText` did not recognize empty lines which made it different than `GetFullWordSpan`. This is fixed now. 
1. Changed `IsWordChar` to `IsKeywordChar` and removed the `WordKind` parameter.    Implementing `IsWordChar` is hard to do correctly because it requires  context. Consider `\n` as an example: is that a word char or not? Cannot  be correctly answered unless you have the surrounding context which the function, by definition, does not. The member  `IsKeywordChar` though has a precise implementation and everyone was passing `WordKind.NormalWord` anyways.

Reply:
Interesting about the missing `Seq,,rev`.. The rationale doesn't convince me because the do provide `Seq.sort` and that also requires linear space to implement. Not providing it means each user has to choose which alternative is the best. It's an interesting topic!

Reply:
Hmm true about `Seq.sort` having the same downside as `Seq.rev`.  Looked into it more and seems there is a lot of passion around keeping `Seq.rev` out of F# but can't find similar conversations around `Seq.sort`. 

> https://github.com/dotnet/fsharp/issues/902